src/main/java/htsjdk/samtools/cram/structure/CramHeader.java

```diff
@@ -24,50 +24,87 @@
 import java.util.Objects;
 
 /**
- * A starting object when dealing with CRAM files. A {@link CramHeader} holds 2 things: 1. File format definition, including content id and
- * version information 2. SAM file header
+ * A starting object when dealing with CRAM files. A {@link CramHeader} holds 2 things:
+ * 1. File format definition, including content id and version information
+ * 2. SAM file header
  */
 public final class CramHeader {
     public static final byte[] MAGIC = "CRAM".getBytes();
+    public static final int MAX_ID_LENGTH = 20;     // in bytes; from CRAM spec
 
     private Version version;
-    private final byte[] id = new byte[20];
-
-    {
-        Arrays.fill(id, (byte) 0);
-    }
+    private byte[] id;
 
     private SAMFileHeader samFileHeader;
 
     /**
-     * Create a new {@link CramHeader} empty object.
+     * Create a new {@link CramHeader} object with the specified version, id and SAM file header.
+     * The id field will be truncated to 20 bytes.
+     *
+     * @param version       the CRAM version to assume
+     * @param id            an identifier of the content associated with this header
+     * @param samFileHeader the SAM file header
      */
-    private CramHeader() {
+    public CramHeader(final Version version, final String id, final SAMFileHeader samFileHeader) {
+        this.version = version;
+        this.samFileHeader = samFileHeader;
+
+        if (id != null) {
+            setId(id);
+        }
     }
 
     /**
      * Create a new {@link CramHeader} object with the specified version, id and SAM file header.
-     * The id field by default is guaranteed to be byte[20].
+     * The id field will be truncated to 20 bytes.
      *
      * @param version       the CRAM version to assume
      * @param id            an identifier of the content associated with this header
      * @param samFileHeader the SAM file header
      */
-    public CramHeader(final Version version, final String id, final SAMFileHeader samFileHeader) {
+    public CramHeader(final Version version, final byte[] id, final SAMFileHeader samFileHeader) {
         this.version = version;
-
-        if (id != null) System.arraycopy(id.getBytes(), 0, this.id, 0, Math.min(id.length(), this.id.length));
         this.samFileHeader = samFileHeader;
+
+        if (id != null) {
+            setId(id);
+        }
+    }
+
+    public byte[] getId() {
+        return id;
+    }
+
+    public String getIdString() {
+        return new String(id);
+    }
+
+    /**
+     * Set the id of the header. A typical use is for example file name to be used when streaming or a checksum of the
+     * data contained in the file.
+     *
+     * @param id a new id; only first 20 bytes will be used.
+     */
+    public void setId(final byte[] id) {
```

Comment:
It looks like `setId` is unused in the code base. Can both versions be removed, and `id` be made final?

Reply:
ok.  but note #1307